stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,187 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+      return Array.TangentVector(
+        elements: zippedElements.map { ($0 ?? .zero) - ($1 ?? .zero) })
+    }
+
+    @inlinable
+    public func moved(
+      along direction: Array.TangentVector
+    ) -> [Element.TangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(
+      from cotangent: Array.CotangentVector
+    ) -> Array.TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return Array.TangentVector(elements: elements)
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias AllDifferentiableVariables = Array.CotangentVector
+    public typealias TangentVector = Array.CotangentVector
+    public typealias CotangentVector = Array.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: Array.CotangentVector {
+      return Array.CotangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (
+      lhs: Array.CotangentVector,
+      rhs: Array.CotangentVector
+    ) -> Array.CotangentVector {
+      let zippedElements = zipLongest(lhs.elements, rhs.elements)
+      return Array.CotangentVector(
+        elements: zippedElements.map { ($0 ?? .zero) + ($1 ?? .zero) })
+    }
+
+    @inlinable
+    public static func - (
+      lhs: Array.CotangentVector,
+      rhs: Array.CotangentVector
+    ) -> Array.CotangentVector {
+      let zippedElements = zipLongest(lhs.elements, rhs.elements)
+      return Array.CotangentVector(
+        elements: zippedElements.map { ($0 ?? .zero) - ($1 ?? .zero) })
+    }
+
+    @inlinable
+    public func moved(
+      along direction: Array.CotangentVector
+    ) -> [Element.CotangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(
+      from cotangent: Array.TangentVector
+    ) -> Array.CotangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return Array.CotangentVector(elements: elements)
+    }
+  }
+
+  @inlinable
+  public func moved(along direction: TangentVector) -> [Element] {
+    return zip(self, direction.elements).map { $0.moved(along: $1) }
+  }
+
+  @inlinable
+  public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+    return TangentVector(elements: zip(self, cotangent.elements).map { $0.tangentVector(from: $1) })
+  }
+}
+
+// TODO: Remove once `zipLongest` is part of stdlib.
+@usableFromInline
+internal func zipLongest<Sequence1: Sequence, Sequence2: Sequence>(
+  _ sequence1: Sequence1, 
+  _ sequence2: Sequence2
+) -> [(Sequence1.Element?, Sequence2.Element?)] {
```

Comment:
I don't think it's ideal to make this return tuples of optionals. This would lose semantic guarantees that the elements in the shortest-zipped region are non-nil, and it is inefficient. Instead, this should take an extra argument that represents the the value used to fill the empty area. Moreover, since we aren't designing a new stdlib API here and that every tangent vector operation takes arguments of the same type, i don't see a need for generalizing this helper function by having more than one generic parameter.

```swift
internal func zipLongest<S: Sequence>(
  _ sequence1: S, 
  _ sequence2: S,
  default: S.Element
) -> [(S.Element, S.Element)] {
  ...
}
```

Reply:
More ideally, this should be made lazy for efficiency, but we can do that later.

Reply:
Good point! I'll change that.